server/src/main/java/org/elasticsearch/cluster/coordination/NodeJoinExecutor.java

```diff
@@ -139,12 +139,22 @@ public ClusterState execute(BatchExecutionContext<JoinTask> batchExecutionContex
                 final DiscoveryNode node = nodeJoinTask.node();
                 if (currentNodes.nodeExistsWithSameRoles(node)) {
                     logger.debug("received a join request for an existing node [{}]", node);
+
+                    // update the node's feature set if it has one
+                    // this can happen if the master has just moved from a pre-features version to a post-features version
```

Comment:
This will eventually become dead code when we drop support for older versions, so deserves a mention of some approprately-old version constant to force its removal.

Reply:
Added an assert check